src/storage-operators/src/oneshot_source.rs

```diff
@@ -0,0 +1,892 @@
+            }
+            .await
+            .context("decode chunk");
+
+            match result {
+                Ok(rows) => rows
+                    .into_iter()
+                    .for_each(|row| row_handle.give(&capability, Ok(row))),
+                Err(err) => row_handle.give(&capability, Err(err)),
+            }
+        }
+    });
+
+    (row_stream, shutdown.press_on_drop())
+}
+
+/// Render an operator that given a stream of [`Row`]s will stage them in Persist and return a
+/// stream of [`ProtoBatch`]es that can later be linked into a shard.
+pub fn render_stage_batches_operator<G>(
+    scope: G,
+    collection_id: GlobalId,
+    collection_meta: &CollectionMetadata,
+    persist_clients: Arc<PersistClientCache>,
+    rows_stream: &TimelyStream<G, Result<Row, StorageErrorX>>,
+) -> (
+    TimelyStream<G, Result<ProtoBatch, StorageErrorX>>,
+    PressOnDropButton,
+)
+where
+    G: Scope,
+{
+    let persist_location = collection_meta.persist_location.clone();
+    let shard_id = collection_meta.data_shard;
+    let collection_desc = collection_meta.relation_desc.clone();
+
+    let mut builder =
+        AsyncOperatorBuilder::new("CopyFrom-stage_batches".to_string(), scope.clone());
+
+    let (proto_batch_handle, proto_batch_stream) =
+        builder.new_output::<CapacityContainerBuilder<_>>();
+    let mut rows_handle = builder.new_input_for(rows_stream, Pipeline, &proto_batch_handle);
+
+    let shutdown = builder.build(move |caps| async move {
+        let [_proto_batch_cap] = caps.try_into().unwrap();
+
+        // Open a Persist handle that we can use to stage a batch.
+        let persist_client = persist_clients
+            .open(persist_location)
+            .await
+            .expect("failed to open Persist client");
+        let persist_diagnostics = Diagnostics {
+            shard_name: collection_id.to_string(),
+            handle_purpose: "CopyFrom::stage_batches".to_string(),
+        };
+        let write_handle = persist_client
+            .open_writer::<SourceData, (), mz_repr::Timestamp, Diff>(
+                shard_id,
+                Arc::new(collection_desc),
+                Arc::new(UnitSchema),
+                persist_diagnostics,
+            )
+            .await
+            .expect("could not open Persist shard");
+
+        // TODO(parkmcar): Use different timestamps here.
+        let lower = mz_repr::Timestamp::MIN;
+        let upper = Antichain::from_elem(lower.step_forward());
+
+        let mut batch_builder = write_handle.builder(Antichain::from_elem(lower));
+        let mut caps = CapabilitySet::new();
+
+        while let Some(event) = rows_handle.next().await {
+            let AsyncEvent::Data(cap, row_batch) = event else {
+                continue;
+            };
+
+            // Note: In the current implementation of `COPY FROM` this set should
+            // only ever contain one capability.
+            caps.insert(cap);
```

Comment:
Is the manual capability handling necessary here? Istm you could just use the initial operator capability instead.

Reply:
Not sure! It felt like this was the most "correct" approach, but originally I did just use the initial operator capability. Would you prefer one over the other?

Reply:
I'd prefer using the initial capability because it's simpler. Also, I think the current approach requires the batch operator to rely on the other operators to provide data capabilities at `Timestamp::MIN`, right? Which is not an unreasonable requirement to have, but using the initial capability lets us not need it, making things a bit easier to reason about.

Reply:
Right! This all makes sense to me, and makes the code cleaner. I removed the use of the `CapabilitySet` and rely on the initial capability of the operator